### ISSUE TITLE
error: implementer accepts interface, not {interface subclass}

### ISSUE THREAD

**Issue by jaraco** [Shoobx/mypy-zope#26]
I think I may have stumbled onto another defect in `mypy.zope`.  Consider this code:

```
from zope.interface import implementer
from twisted.cred.credentials import ICredentials


class IToken(ICredentials):
    pass


@implementer(IToken)
class Token:
    pass

```

Run that through mypy with this mypy.ini:

```
[mypy]
ignore_missing_imports = True
plugins=mypy_zope:plugin
```

With twisted and zope interface installed, and this error occurs:

```
$ .tox/typechecks/bin/python -m mypy --no-incremental creds.py
creds.py:9: error: zope.interface.implementer accepts interface, not creds.IToken.
creds.py:9: error: Make sure you have stubs for all packages that provide interfaces for creds.IToken class hierarchy.
Found 2 errors in 1 file (checked 1 source file)
```

But make either of these tweaks and the errors go away:

1. Save the implementer in a separate variable.

```
from zope.interface import implementer
from twisted.cred.credentials import ICredentials


class IToken(ICredentials):
    pass


_impl = implementer(IToken)

@_impl
class Token:
    pass
```

2. Inline the definition of ICredentials.

```
from zope.interface import implementer, Interface


class ICredentials(Interface):
    pass


class IToken(ICredentials):
    pass


@implementer(IToken)
class Token:
    pass

```

**Comment by jaraco** [comment]
I welcome any advice that explains the error and recommends a workaround for this piece of code.

**Comment by kedder** [comment]
The problem (I think) is that `ICredentials` from `twisted.cred.credentials` is not recognized to be descendant of Interface. This is because (I suspect) twisted does not have typing stubs, so there is no chance for mypy to know the actual type of the ICredentials. 

The remedy is to use `stubgen` (that comes with mypy package) to generate stubs for twisted package, and to include them in your `mypy.ini`: 

```
[mypy]
...
plugins=mypy_zope:plugin
mypy_path=stubs
```

(assuming stubs are placed in `stubs` directory).

**Comment by jaraco** [comment]
Thanks. I'll give that a try.

**Comment by jaraco** [comment]
I've generated the stubs from twisted trunk and pushed them to the target repo, but I notice this creates a new error:

```
tahoe-lafs 3399.mypy $ .tox/typechecks/bin/python -m mypy --no-incremental creds.py
stubs/twisted/python/deprecate.pyi:30: error: Name '_Tc' is not defined
Found 1 error in 1 file (checked 1 source file)
```

I guessed this outcome was because of a mismatch between the version of twisted, so I re-generated the stubs using the version of twisted as found in the tahoe-lafs project, but then I got different errors, seemingly related to latent Python 2 code:

```
tahoe-lafs 3399.mypy $ .tox/typechecks/bin/python -m mypy --no-incremental creds.py
stubs/twisted/python/compat.pyi:3: error: Module 'base64' has no attribute 'decodestring'
stubs/twisted/python/compat.pyi:3: error: Module 'base64' has no attribute 'encodestring'
stubs/twisted/python/compat.pyi:4: error: Module 'cgi' has no attribute 'escape'
stubs/twisted/python/compat.pyi:9: error: Module 'types' has no attribute 'FileType'
stubs/twisted/python/compat.pyi:9: error: Module 'types' has no attribute 'InstanceType'
stubs/twisted/python/compat.pyi:11: error: Module 'urllib' has no attribute 'quote'
stubs/twisted/python/compat.pyi:11: error: Module 'urllib' has no attribute 'unquote'
stubs/twisted/python/compat.pyi:19: error: Name 'cmp' is not defined
stubs/twisted/python/compat.pyi:33: error: Name 'buffer' is not defined
stubs/twisted/python/compat.pyi:34: error: Name 'basestring' is not defined
stubs/twisted/python/compat.pyi:45: error: Cannot assign to a type
Found 11 errors in 1 file (checked 1 source file)
```

I'll continue to investigate.

**Comment by kedder** [comment]
The stubs generated by `stubgen` are more like a starting point, they usually need adjustments and tweaks. Sometimes the runtime magic in some modules does not translate well to stub files. So, take a look where that `_Tc` comes from, maybe you just need to delete some lines in the generated stub. Also, since .pyi files are not executable and you are likely don't care about py2 compatibility, you can just delete the `compat.pyi` from your stubs.

**Comment by jaraco** [comment]
Good tip. Yeah, I tried deleting the compat.pyi and the error goes away for the minimal reproducer \o/. Unfortunately, as you've indicated might happen, the generated stubs lead to hundreds of new errors when included in the Tahoe suite (104 without stubs, 464 with). At least, I think we've confirmed the _proper_ fix for this issue, which doesn't implicate mypy-zope at all, so I'll close this issue. Thanks for all the help.
